service/storage_service.cc

```diff
@@ -2015,13 +2018,27 @@ future<> storage_service::decommission() {
                     heartbeat_updater.get();
                 });
 
-                // Step 5: Start to sync data
+                // Step 4: Start to sync data
                 slogger.info("DECOMMISSIONING: unbootstrap starts");
                 ss.unbootstrap().get();
-                ss.leave_ring().get();
                 slogger.info("DECOMMISSIONING: unbootstrap done");
 
-                // Step 6: Finish
+                // Step 5: Become a group 0 non-voter before leaving the token ring.
+                //
+                // Thanks to this, even if we fail after leaving the token ring but before leaving group 0,
+                // group 0's availability won't be reduced.
+                if (raft_available) {
+                    slogger.info("decommission[{}]: becoming a group 0 non-voter", uuid);
+                    ss._group0->become_nonvoter().get();
+                    slogger.info("decommission[{}]: became a group 0 non-voter", uuid);
+                }
+
```

Comment:
If we crash at this step, can decommission get reverted and we stay a non-voter forever?

Reply:
If we crash here, the user can restart the node, the node will still be part of the ring. Will the raft code during restart, make the node a voter again?

Reply:
The node will become a voter again during restart, yes.

Reply:
Good.

Reply:
There is another tool we can use to revert the non-voter to voter in case of failure. 

There is a abort function attached to each node_ops_meta_data. The abort function will be called automatically if the decommission node does not update in a timrly manager (like a watchdog) to revert the decommission ops.

Perhaps we can call 'add the decommission node to a voter' in the abort function. This way, the decommission node will become voter automatically without user's operation. 

But on the other hand, without restarting the decommission node, it does not really matter if the decommission node is a voter or a non-voter, since we need to restart the decommission node to make it work properly again anyway. 

```
            auto meta = node_ops_meta_data(ops_uuid, coordinator, std::move(req.ignore_nodes), [this, coordinator, req = std::move(req)] () mutable {
                return mutate_token_metadata([this, coordinator, req = std::move(req)] (mutable_token_metadata_ptr tmptr) mutable {
                    for (auto& node : req.leaving_nodes) {
                        slogger.info("decommission[{}]: Removed node={} as leaving node, coordinator={}", req.ops_uuid, node, coordinator);
                        tmptr->del_leaving_endpoint(node);
                    }
                    return update_pending_ranges(tmptr, format("decommission {}", req.leaving_nodes)); 
                }); 
            },
            [this, ops_uuid] () mutable { node_ops_singal_abort(ops_uuid); });
            _node_ops.emplace(ops_uuid, std::move(meta));
```

Reply:
> But on the other hand, without restarting the decommission node, it does not really matter if the decommission node is a voter or a non-voter, since we need to restart the decommission node to make it work properly again anyway.


This.

And it can actually be considered a good thing that it is a non-voter while it is down.